File: engine_py/tests/test_loot.py
"""Tests for the Python loot table API bindings."""


def test_define_and_roll(make_world):
    world = make_world()
    result = world.define_loot_table("test", [
        {"item_id": "health_potion", "weight": 100}
    ])
    assert result is True
    drops = world.roll_loot_table("test")
    assert len(drops) == 1
    assert drops[0]["item_id"] == "health_potion"
    assert drops[0]["count"] == 1


def test_define_with_min_max_count(make_world):
    world = make_world()
    result = world.define_loot_table("multi", [
        {"item_id": "coins", "weight": 100, "min_count": 2, "max_count": 5}
    ])
    assert result is True
    for _ in range(20):
        drops = world.roll_loot_table("multi")
        assert len(drops) == 1
        assert 2 <= drops[0]["count"] <= 5


def test_multiple_entries(make_world):
    world = make_world()
    world.define_loot_table("mixed", [
        {"item_id": "common", "weight": 90},
        {"item_id": "rare", "weight": 10},
    ])
    drops = world.roll_loot_table("mixed")
    assert 1 <= len(drops) <= 2


def test_has_loot_table(make_world):
    world = make_world()
    assert world.has_loot_table("nonexistent") is False
    world.define_loot_table("foo", [{"item_id": "bar", "weight": 100}])
    assert world.has_loot_table("foo") is True


def test_loot_table_names(make_world):
    world = make_world()
    world.define_loot_table("a", [{"item_id": "x", "weight": 100}])
    world.define_loot_table("b", [{"item_id": "y", "weight": 100}])
    names = world.loot_table_names()
    assert len(names) == 2
    assert "a" in names
    assert "b" in names


def test_remove_loot_table(make_world):
    world = make_world()
    world.define_loot_table("temp", [{"item_id": "x", "weight": 100}])
    assert world.has_loot_table("temp") is True
    world.remove_loot_table("temp")
    assert world.has_loot_table("temp") is False


def test_roll_nonexistent_table(make_world):
    world = make_world()
    import pytest
    with pytest.raises(ValueError, match="not found"):
        world.roll_loot_table("nonexistent")


def test_define_zero_weight_raises(make_world):
    world = make_world()
    import pytest
    with pytest.raises(ValueError, match="zero weight"):
        world.define_loot_table("bad", [
            {"item_id": "item", "weight": 0}
        ])


def test_define_min_gt_max_raises(make_world):
    world = make_world()
    import pytest
    with pytest.raises(ValueError):
        world.define_loot_table("bad", [
            {"item_id": "x", "weight": 100, "min_count": 5, "max_count": 1}
        ])


def test_define_overwrites(make_world):
    world = make_world()
    world.define_loot_table("dupe", [{"item_id": "old", "weight": 100}])
    world.define_loot_table("dupe", [{"item_id": "new", "weight": 100}])
    drops = world.roll_loot_table("dupe")
    assert drops[0]["item_id"] == "new"


def test_empty_table(make_world):
    world = make_world()
    world.define_loot_table("empty", [])
    import pytest
    with pytest.raises(ValueError, match="no entries"):
        world.roll_loot_table("empty")


def test_missing_item_id_raises(make_world):
    world = make_world()
    import pytest
    with pytest.raises((ValueError, TypeError)):
        world.define_loot_table("bad", [{"weight": 100}])

File: engine/scripts/lua/tests/test_loot.lua
-- test_loot.lua: Tests for the Rust-native loot table API.
-- Each test gets a fresh world via the test runner.
-- Global functions: define_loot_table(), roll_loot_table(), has_loot_table()

local assert = require("assert")

-- 1. Define table with 100% weight and roll returns items
local function test_define_and_roll_basic()
    local ok = define_loot_table("test_basic", {
        { item_id = "health_potion", weight = 100 },
    })
    assert.is_true(ok, "define_loot_table should return true")
    local results = roll_loot_table("test_basic")
    assert.is_true(#results > 0, "Should roll at least one item")
    assert.equals(results[1].item_id, "health_potion")
    assert.equals(results[1].count, 1)
end

-- 2. Rolling undefined table returns empty table without error
local function test_undefined_table_returns_empty()
    local results = roll_loot_table("nonexistent_table")
    assert.is_table(results)
    assert.equals(#results, 0)
end

-- 3. Multiple entries: weight=100 items always drop
local function test_multiple_entries_resolve()
    define_loot_table("test_multi", {
        { item_id = "health_potion", weight = 100 },
        { item_id = "rusty_sword", weight = 100 },
    })
    local results = roll_loot_table("test_multi")
    -- Both have 100% weight, so both should always drop
    assert.is_true(#results == 2, "Both entries should drop")
    local ids = {}
    for _, drop in ipairs(results) do
        ids[drop.item_id] = true
    end
    assert.is_true(ids["health_potion"], "health_potion should appear")
    assert.is_true(ids["rusty_sword"], "rusty_sword should appear")
end

-- 4. Weighted distribution: both items appear over many rolls
local function test_weighted_distribution()
    define_loot_table("test_weighted", {
        { item_id = "health_potion", weight = 90 },
        { item_id = "rusty_sword", weight = 10 },
    })
    local seen = {}
    for _ = 1, 100 do
        local results = roll_loot_table("test_weighted")
        for _, drop in ipairs(results) do
            seen[drop.item_id] = true
        end
    end
    assert.is_true(seen["health_potion"], "health_potion should appear in 100 rolls")
    assert.is_true(seen["rusty_sword"], "rusty_sword should appear in 100 rolls")
end

-- 5. min_count/max_count: count stays within range
local function test_min_max_count()
    define_loot_table("test_count", {
        { item_id = "coins", weight = 100, min_count = 2, max_count = 5 },
    })
    for _ = 1, 20 do
        local results = roll_loot_table("test_count")
        assert.is_true(#results >= 1, "Should drop at least one entry")
        if #results > 0 then
            local count = results[1].count
            assert.is_true(count >= 2 and count <= 5,
                "Count " .. count .. " should be in [2, 5]")
        end
    end
end

-- 6. Redefining a table overwrites previous entries
local function test_redefine_table_overwrites()
    define_loot_table("overwrite_test", {
        { item_id = "health_potion", weight = 100 },
    })
    define_loot_table("overwrite_test", {
        { item_id = "rusty_sword", weight = 100 },
    })
    local results = roll_loot_table("overwrite_test")
    assert.is_true(#results > 0)
    assert.equals(results[1].item_id, "rusty_sword",
        "Redefined table should drop rusty_sword")
end

-- 7. has_loot_table returns correct booleans
local function test_has_loot_table()
    assert.is_false(has_loot_table("unknown"), "Unknown table should return false")
    define_loot_table("known_table", {
        { item_id = "test", weight = 100 },
    })
    assert.is_true(has_loot_table("known_table"), "Defined table should return true")
end

-- 8. define_loot_table rejects zero-weight entries
local function test_zero_weight_rejected()
    local ok, err = pcall(define_loot_table, "bad", {
        { item_id = "test", weight = 0 },
    })
    assert.is_false(ok, "Zero-weight entry should be rejected")
end

-- 9. Define empty entries table succeeds, but rolling returns empty
local function test_empty_table_rolls_empty()
    local ok = define_loot_table("empty", {})
    assert.is_true(ok, "Empty entries table should be accepted")
    local results = roll_loot_table("empty")
    assert.is_table(results)
    assert.equals(#results, 0, "Rolling empty table should return empty")
end

-- 10. Multiple items from a single entry (min_count > 1)
local function test_multi_spawn_from_single_entry()
    define_loot_table("multi_spawn", {
        { item_id = "arrows", weight = 100, min_count = 3, max_count = 3 },
    })
    local results = roll_loot_table("multi_spawn")
    assert.is_true(#results >= 1, "Should drop at least one entry")
    if #results > 0 then
        assert.equals(results[1].item_id, "arrows")
        assert.equals(results[1].count, 3)
    end
end

return {
    test_define_and_roll_basic = test_define_and_roll_basic,
    test_undefined_table_returns_empty = test_undefined_table_returns_empty,
    test_multiple_entries_resolve = test_multiple_entries_resolve,
    test_weighted_distribution = test_weighted_distribution,
    test_min_max_count = test_min_max_count,
    test_redefine_table_overwrites = test_redefine_table_overwrites,
    test_has_loot_table = test_has_loot_table,
    test_zero_weight_rejected = test_zero_weight_rejected,
    test_empty_table_rolls_empty = test_empty_table_rolls_empty,
    test_multi_spawn_from_single_entry = test_multi_spawn_from_single_entry,
}
